core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1337,290 +1451,127 @@ class Block : public lifecycle::StateMachine<Derived>, //
-                }
-                publishTag({ { gr::tag::END_OF_STREAM, true } }, 0);
-                return { requested_work, ports_status.in_samples, success ? work::Status::DONE : work::Status::ERROR };
+        } else { // block does not define any valid processing function
+            static_assert(gr::meta::always_false<gr::traits::block::stream_input_port_types_tuple<Derived>>, "neither processBulk(...) nor processOne(...) implemented");
+        }
+        forwardTags();
+        if (lifecycle::isShuttingDown(this->state()) || nextEosTag <= processedIn + 1) {
+            if (auto e = this->changeStateTo(lifecycle::State::REQUESTED_STOP); !e) {
+                emitErrorMessage("isShuttingDown -> STOPPED", e.error());
             }
-
-            // return { requested_work, ports_status.in_samples, success ? (lifecycle::isShuttingDown(std::atomic_load_explicit(&state, std::memory_order_acquire)) ? work::Status::DONE :
-            // work::Status::OK) : work::Status::ERROR };
-            return { requested_work, ports_status.in_samples, success ? work::Status::OK : work::Status::ERROR };
-        } // processOne(...) handling
-        else {
-            static_assert(gr::meta::always_false<Derived>, "neither processBulk(...) nor processOne(...) implemented");
+            updateInputAndOutputTags(processedIn);
+            applyChangedSettings();
+            ret       = work::Status::DONE;
+            processedIn = 0UZ;
         }
-        return { requested_work, 0UZ, work::Status::ERROR };
-    } // end: work_return_t work_internal() noexcept { ..}
+        // publish/consume
+        publishSamples(processedOut, outputSpans);
+        const auto inputSamplesToSkipAfter = resampledIn > 0 ? inputSamplesToSkipAfterChunk(maxSyncAvailableIn - inputSkipBefore) : 0UZ;
+        bool       success;
+        if (inputSamplesToSkipAfter > 0 && inputSamplesToSkipAfter <= resampledIn) {
+            updateInputAndOutputTags(inputSamplesToSkipAfter); // apply all tags in the skipped data range
+            success = consumeReaders(inputSamplesToSkipAfter, inputSpans);
+        } else if (inputSamplesToSkipAfter > 0) {
+            updateInputAndOutputTags(inputSamplesToSkipAfter); // apply all tags in the skipped data range
+            success                          = consumeReaders(resampledIn, inputSpans);
+            const auto inputSpansForSkipping = prepareStreams(inputPorts<PortType::STREAM>(&self()), inputSamplesToSkipAfter - resampledIn); // only way to consume is via the ConsumableSpan now
+            consumeReaders(inputSamplesToSkipAfter - resampledIn, inputSpansForSkipping);
+        } else {
+            success = consumeReaders(processedIn, inputSpans);
+        }
+        // if the block state changed to DONE, publish EOS tag on the next sample
+        if (ret == work::Status::DONE) {
+            this->setAndNotifyState(lifecycle::State::STOPPED);
+            publishTag({ { gr::tag::END_OF_STREAM, true } }, 1);
```

Comment:
Why index 1 here, and not 0 (like in other places?)?

Reply:
The done tag always gets the index of the next sample (which will never be published). Otherwise a block could not publish EoS without publishing more samples. But you are right, that seems to be already considered in publishTag, it does not publish to the last written but to the next written index. Probably it doesn't matter because in the next iteration it sees the state and publishes another tag on the correct position... so maybe it should just be removed s.t. all eos tag handling is done at the beginning of the function...